core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3708,26 +3708,66 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
    */
   def fromFuture[A](make: ExecutionContext => scala.concurrent.Future[A])(implicit trace: Trace): Task[A] =
     ZIO.executorWith { executor =>
-      import scala.util.{Failure, Success}
-      val ec = executor.asExecutionContext
-      ZIO.attempt(make(ec)).flatMap { f =>
-        val canceler: UIO[Unit] = f match {
-          case cancelable: CancelableFuture[A] =>
-            ZIO.suspendSucceed(if (f.isCompleted) ZIO.unit else ZIO.fromFuture(_ => cancelable.cancel()).ignore)
-          case _ => ZIO.unit
-        }
+      ZIO.suspend {
+        import scala.util.{Success, Failure}
 
-        f.value
-          .fold(
+        val ec = executor.asExecutionContext
+        val f  = make(ec)
+        f.value match {
+          case None =>
             ZIO.asyncInterrupt { (k: Task[A] => Unit) =>
               f.onComplete {
                 case Success(a) => k(Exit.succeed(a))
                 case Failure(t) => k(ZIO.fail(t))
               }(ec)
 
-              Left(canceler)
+              Left {
+                f match {
+                  case cf: CancelableFuture[A] =>
+                    ZIO.suspendSucceed(if (cf.isCompleted) Exit.unit else ZIO.fromFuture(cf.cancel()).ignore)
+                  case _ =>
+                    Exit.unit
+                }
+              }
+            }
+          case Some(Success(value)) => Exit.succeed(value)
+          case Some(Failure(t))     => ZIO.fail(t)
+        }
+      }
+    }
+
+  /**
+   * Returns an effect that, when executed, will both create and launch a
+   * [[scala.concurrent.Future]] and run it on ZIO's own executor.
+   *
+   * @see
+   *   Overloaded method that allows using the ZIO executor-backed execution
+   *   context
+   */
+  def fromFuture[A](future: => scala.concurrent.Future[A])(implicit trace: Trace): Task[A] =
```

Comment:
Seems okay - slightly concerning that people may accidentally use a distinct EC due to the signature.

Reply:
Hmm if we think it's better not to expose this to users, I can always make it package-private. Wdut?

Reply:
Sounds good to me. How would you feel about dropping the by name parameter? We can avoid an allocation if it's only internal?

Reply:
Sounds good to me